text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,533 @@
+
+If we add all of these with the current design, that would increase the size of
+`aws-cdk-lib` by approximately 130 MiB*, and make the Lambda Layers the biggest
+contributor of size at 39%.
+
+*See Appendix A to see where this estimate came from.
+
+| Category                       | Percentage of aws-cdk-lib size |
+| ------------------------------ | ------------------------------ |
+| Lambda Layer zip files         | 38.83%                         |
+| .jsii files                    | 30.14%                         |
+| Source Maps (.js.map)          | 15.87%                         |
+| Javascript code (.js)          | 6.63%                          |
+| Type Declaration files (.d.ts) | 6.62%                          |
+| README, etc (.md)              | 0.55%                          |
+| bundled npm dependencies       | 0.50%                          |
+| .ts-fixture                    | 0.27%                          |
+| .jsiirc                        | 0.26%                          |
+
+### Why should we _not_ do this?
+
+The main reason to not do this is that it will inevitably break up the
+monolothic `aws-cdk-lib` package in some manner, and we will lose some of the
+wonderful simplicity of releasing and using a single package. But, we can't keep
+everything in one place, and reduce the size enough to have a significant
+positive impact.
+
+### What is the technical solution (design) of this feature?
+
+There are different categories of files contributing to the size of aws-cdk-lib,
+and each category will have it’s own solution. The following sections cover the
+top five categories in size contribution in order from largest to smallest
+potential size.
+
+#### Lambda Layer Zip Files
+
+Use separate npm packages for each `lambda-layer-X` module, which the CDK CLI
+will automatically install if not already present in the `node_modules`
+directory. With this solution, customers will not have to figure out when they
+do or do not need to include these dependencies, and they will not be required
+to access a public endpoint besides npm during runtime of their CDK apps.
+Practically, this solution breaks down into the following steps.
+
+1. Publish v2 compatible versions of `@aws-cdk/lambda-layer-aws-cli`,
+   `@aws-cdk/lambda-layer-aws-cli`, `@aws-cdk/lambda-layer-node-proxy-agent` as
+   their own npm packages, separate from `aws-cdk-lib`.
+2. Modify the `lambda-layer-X` submodules in `aws-cdk-lib` to not bundle these
+   large dependencies themselves, but instead reference the appropriate `Layer`
+   class from the packages in step 1. We will need to do something clever here
+   to make sure customers do not have IDE or compilation errors before the CDK
+   CLI has an opportunity to download and install the lambda layer packages.
+3. Modify the CDK CLI to verify that the correct packages are available in the
+   `node_modules` directory during synthesis. If they are not available,
+   download them from npm.
```

Comment:
How does this work for non-NodeJS languages, that might not have a `node_modules` directory at all?

Reply:
`~/.cache/cdk/large-packages` ?

Reply:
They still need to have a global `node_modules` directory to have the CDK CLI installed. I think we can attempt to install into the current application's `node_modules`, if there isn't one, then install into the global one. Although, will elevated permissions be required for that?

Reply:
> I think we can attempt to install into the current application's `node_modules`, if there isn't one, then install into the global one. Although, will elevated permissions be required for that?

Also, won't that cause problems if the customer has multiple CDK apps with different versions on their machine?

Reply:
Why not create an application level `node_modules` if it doesn't exist?

Reply:
The proposal has changed a bit since this conversation. The dependencies will be downloaded at runtime from the framework, not the CLI. Check out the latest version, and let me know if there is any further concerns about this topic!

Reply:
Love the new direction!